core/src/databases/csv.rs

```diff
@@ -0,0 +1,323 @@
+use anyhow::{anyhow, Result};
+use cloud_storage::Object;
+use csv_async::AsyncReaderBuilder;
+use futures::stream::StreamExt;
+use lazy_static::lazy_static;
+use tokio::io::AsyncReadExt;
+
+use regex::Regex;
+use tokio_util::compat::TokioAsyncReadCompatExt;
+use unicode_normalization::UnicodeNormalization;
+
+use crate::databases::table::Row;
+
+pub struct UpsertQueueCSVContent {
+    pub upsert_queue_bucket_path: String,
+}
+
+const MAX_TABLE_COLUMNS: usize = 512;
+const MAX_COLUMN_NAME_LENGTH: usize = 1024;
+
+impl UpsertQueueCSVContent {
+    async fn get_upsert_queue_bucket() -> Result<String> {
+        match std::env::var("DUST_UPSERT_QUEUE_BUCKET") {
+            Ok(bucket) => Ok(bucket),
+            Err(_) => Err(anyhow!("DUST_UPSERT_QUEUE_BUCKET is not set")),
+        }
+    }
+
+    pub async fn parse(&self) -> Result<Vec<Row>> {
+        let bucket = Self::get_upsert_queue_bucket().await?;
+        let path = &self.upsert_queue_bucket_path;
+
+        // This is not the most efficient as we download the entire file here but we will
+        // materialize it in memory as Vec<Row> anyway so that's not a massive difference.
+        let content = Object::download(&bucket, path).await?;
+        let rdr = std::io::Cursor::new(content);
+
+        let (delimiter, rdr) = Self::find_delimiter(rdr).await?;
+        Self::csv_to_rows(rdr, delimiter).await
+    }
+
+    fn slugify(text: &str) -> String {
+        lazy_static! {
+            static ref DIACRITICS: Regex = Regex::new(r"[\u{0300}-\u{036f}]").unwrap();
+        }
+
+        let s = text
+            .nfkd() // Normalize to decomposed form
+            .collect::<String>();
+        let s = DIACRITICS.replace_all(&s, "").to_string();
+
+        // Insert _ between lowercase and uppercase/number
+        let mut with_separators = String::new();
+        let chars: Vec<char> = s.chars().collect();
+        for (i, &c) in chars.iter().enumerate() {
+            if i > 0 && c.is_uppercase() || c.is_ascii_digit() {
+                if chars[i - 1].is_lowercase() {
+                    with_separators.push('_');
+                }
+            }
+            with_separators.push(c);
+        }
+
+        with_separators
+            .to_lowercase()
+            .trim()
+            .split_whitespace()
+            .collect::<Vec<_>>()
+            .join("_") // Replace spaces with _
+            .chars()
+            .map(|c| if c.is_alphanumeric() { c } else { '_' }) // Replace non-word chars
+            .collect::<String>()
+            .replace("__", "_") // Replace multiple _ with single _
```

Comment:
This doesn't work if there's more than  2 consecutive `_` no ?

Reply:
written by AI will test and fix

Reply:
(done)